src/lib/gadgets/sha256.ts

```diff
@@ -0,0 +1,278 @@
+// https://csrc.nist.gov/pubs/fips/180-4/upd1/final
+import { mod } from '../../bindings/crypto/finite_field.js';
+import { Field } from '../core.js';
+import { UInt32, UInt8 } from '../int.js';
+import { FlexibleBytes } from '../provable-types/bytes.js';
+import { Bytes } from '../provable-types/provable-types.js';
+import { chunk } from '../util/arrays.js';
+import { TupleN } from '../util/types.js';
+import { divMod32 } from './arithmetic.js';
+import { bytesToWord, wordToBytes } from './bit-slices.js';
+import { bitSlice, exists } from './common.js';
+import { rangeCheck16 } from './range-check.js';
+
+export { SHA256 };
+
+const SHA256Constants = {
+  // constants §4.2.2
+  K: [
+    0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1,
+    0x923f82a4, 0xab1c5ed5, 0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
+    0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174, 0xe49b69c1, 0xefbe4786,
+    0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
+    0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147,
+    0x06ca6351, 0x14292967, 0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
+    0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85, 0xa2bfe8a1, 0xa81a664b,
+    0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
+    0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a,
+    0x5b9cca4f, 0x682e6ff3, 0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
+    0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
+  ],
+  // initial hash values §5.3.3
+  H: [
+    0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c,
+    0x1f83d9ab, 0x5be0cd19,
+  ],
+};
+
+function padding(data: FlexibleBytes): UInt32[][] {
+  // create a provable Bytes instance from the input data
+  // the Bytes class will be static sized according to the length of the input data
+  let message = Bytes.from(data);
+
+  // now pad the data to reach the format expected by sha256
+  // pad 1 bit, followed by k zero bits where k is the smallest non-negative solution to
+  // l + 1 + k = 448 mod 512
+  // then append a 64bit block containing the length of the original message in bits
+
+  let l = message.length * 8; // length in bits
+  let k = Number(mod(448n - (BigInt(l) + 1n), 512n));
+
+  let paddingBits = (
+    '1' + // append 1 bit
+    '0'.repeat(k) + // append k zero bits
+    '0'.repeat(64 - l.toString(2).length) + // append 64bit containing the length of the original message
+    l.toString(2)
+  ).match(/.{1,8}/g)!; // this should always be devisable by 8
+
+  // map the padding bit string to UInt8 elements
+  let padding = paddingBits.map((x) => UInt8.from(BigInt('0b' + x)));
+
+  // concatenate the padding with the original padded data
```

Comment:
the data is unpadded at this point, right? Or  actually padded?

Reply:
yes, the original data is unpadded but will then be concantenated with the padding